File: leetcode/47.permutations-ii.py
# @leet start
class Solution:
    def permuteUnique(self, nums: list[int]) -> list[list[int]]:
        # Store results in a list
        results = []

        # Define recursive function to generate permutations
        def recurse(nums_to_add: list, accum: list) -> None:
            # Base case
            if not nums_to_add:
                results.append(tuple(accum))

            # Generate permutations
            for i, x in enumerate(nums_to_add):
                recurse(nums_to_add[:i] + nums_to_add[i + 1 :], accum + [x])

        # Return all permutations
        recurse(nums, [])

        return [list(perm_tuple) for perm_tuple in set(results)]


# @leet end

File: leetcode/47.permutations-ii.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function (nums) {
  const res = [];

  const helper = (i) => {
    if (i == nums.length) {
      res.push([...nums]);
    }

    const seen = new Set();

    for (let j = i; j < nums.length; j++) {
      if (seen.has(nums[j])) {
        continue;
      }

      seen.add(nums[j]);

      [nums[i], nums[j]] = [nums[j], nums[i]];
      helper(i + 1);
      [nums[i], nums[j]] = [nums[j], nums[i]];
    }
  };

  helper(0);

  return res;
};
// @leet end
